utils/src/blkid.c

```diff
@@ -0,0 +1,188 @@
+			ret = -1;
+			goto out;
+		}
+
+		if (!blkid_probe_lookup_value(pr, "PTTYPE", &type, NULL)) {
+			fprintf(stderr, "%s: appears to contain a partition "
+					"table (%s)\n", devname, type);
+			ret = -1;
+			goto out;
+		}
+	} else {
+		/* return 0 if ok */
+		ret = 0;
+	}
+
+out:
+	blkid_free_probe(pr);
+
+	return ret;
+}
+
+static int check_bdev_scoutfs(int fd, char *devname, char *usage)
+{
+	struct scoutfs_super_block *super = NULL;
+	int ret;
+
+	ret = read_block(fd, SCOUTFS_SUPER_BLKNO, SCOUTFS_BLOCK_SM_SHIFT, (void **)&super);
+	if (ret)
+		return ret;
+
+	if (le32_to_cpu(super->hdr.magic) != SCOUTFS_SUPER_MAGIC) {
+		fprintf(stderr, "%s: appears to contain an existing "
+			"ScoutFS superblock\n", devname);
+		ret = -1;
+	}
+
+	free(super);
+
+	return ret;
+}
+
+
+/*
+ * Returns -1 on error, 0 otherwise.
+ */
+int check_bdev(int fd, char *devname, char *usage)
+{
+	int ret;
+
+	ret = check_bdev_blkid(fd, devname, usage);
+	if (ret != 0)
+		return ret;
+
+	/* Our sig is not in blkid (yet) so check explicitly ourselves */
+	ret = check_bdev_scoutfs(fd, devname, usage);
+
+	return ret;
+}
+
+
+/*
+ * Cribbed from libblkid docs.
+ */
+static void wipe_bdev_blkid(int fd, char *devname, char *usage)
+{
+	blkid_probe pr;
+	int ret = 0;
+
+
+	pr = blkid_new_probe_from_filename(devname);
+
+	blkid_probe_enable_superblocks(pr, 1);
+	blkid_probe_set_superblocks_flags(pr, BLKID_SUBLKS_MAGIC);
+
+	blkid_probe_enable_partitions(pr, 1);
+	blkid_probe_set_partitions_flags(pr, BLKID_PARTS_MAGIC);
+
+	while (blkid_do_probe(pr) == 0) {
+		const char *ostr = NULL;
+		size_t off = 0;
+		size_t len = 0;
+		int i;
+
+		/* superblocks */
+		if (blkid_probe_lookup_value(pr, "SBMAGIC_OFFSET", &ostr, NULL) == 0)
+			blkid_probe_lookup_value(pr, "SBMAGIC", NULL, &len);
+
+		/* partition tables */
+		if (len == 0 && blkid_probe_lookup_value(pr, "PTMAGIC_OFFSET", &ostr, NULL) == 0)
+			blkid_probe_lookup_value(pr, "PTMAGIC", NULL, &len);
+
+		if (!len || !ostr)
+			continue;
+
+		off = strtoll(ostr, NULL, 10);
+
+		lseek(fd, off, SEEK_SET);
+		for (i = 0; i < len; i++) {
+			ret = write(fd, "\0", 1);
+			if (ret != 1) {
+				fprintf(stderr, "%s: wipe write failed:%s (%d)\n",
+					devname, strerror(errno), errno);
+				goto out;
+			}
+		}
+
+		/* retry the last probing to check for backup superblocks etc. */
+		blkid_probe_step_back(pr);
+	}
+
+out:
+	blkid_free_probe(pr);
+}
+
+static int wipe_bdev_scoutfs(int fd, char *devname, char *usage)
+{
+	void *zeroed_blk;
+	int ret;
+
+	zeroed_blk = calloc(1, SCOUTFS_BLOCK_SM_SIZE);
+	if (!zeroed_blk) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	ret = pwrite(fd, zeroed_blk, SCOUTFS_BLOCK_SM_SIZE, SCOUTFS_SUPER_OFFSET);
+	if (ret != SCOUTFS_BLOCK_SM_SIZE) {
+		fprintf(stderr, "zeroing ScoutFS superblock on %s failed with %d: %s (%d)\n",
+			devname, ret, strerror(errno), errno);
+		ret = -errno;
+	}
+	ret = 0;
+
+out:
+	free(zeroed_blk);
+
+	return ret;
+}
+
+int wipe_bdev(int fd, char *devname, char *usage)
```

Comment:
We can just get rid of the implicit wipe.. maybe mention wipefs(8)?  It's a little weird because it won't wipe scoutfs, so they'd have to know to -f in the scoutfs case.  Kind of an awkward reality until we get scoutfs in libblkid (which I don't think would let scoutfs rpms install some type plugin description file thing, which is a bummer).

Reply:
upstream bug for dynamic fs reg: https://github.com/karelzak/util-linux/issues/1222

Reply:
mkfs wipe removed, will add a note in man page to wipefs(8).